stan/math/rev/mat/functor/ode_model.hpp

```diff
@@ -0,0 +1,165 @@
+#ifndef STAN_MATH_REV_MAT_FUNCTOR_ODE_MODEL_HPP
+#define STAN_MATH_REV_MAT_FUNCTOR_ODE_MODEL_HPP
+
+#include <stan/math/rev/core.hpp>
+#include <iostream>
+#include <vector>
+
+namespace stan {
+  namespace math {
+
+    /**
+     * Internal representation of ODE model object which provides
+     * convenient Jacobian functions to obtain gradients wrt to states
+     * (S) and parameters (P). Can be used to provide analytic
+     * Jacobians via partial template specialisation.
+     *
+     * @tparam F type of functor for the base ode system.
+     */
+    template<typename F>
+    struct ode_model {
+      const F& f_;
+      const std::vector<double>& theta_;
+      const std::vector<double>& x_;
+      const std::vector<int>& x_int_;
+      std::ostream* msgs_;
+
+      /**
+       * Construct an ODE model with the specified base ODE system,
+       * parameters, data, and a message stream.
+       *
+       * @param[in] f the base ODE system functor.
+       * @param[in] theta parameters of the ode.
+       * @param[in] x real data.
+       * @param[in] x_int integer data.
+       * @param[in, out] msgs stream to which messages are printed.
+       */
+      ode_model(const F& f,
+                const std::vector<double>& theta,
+                const std::vector<double>& x,
+                const std::vector<int>& x_int,
+                std::ostream* msgs)
+        : f_(f),
+          theta_(theta),
+          x_(x),
+          x_int_(x_int),
+          msgs_(msgs)
+      {}
+
+      /**
+       * Calculate the RHS of the ODE
+       *
+       * @param[in] y state of the ode system at time t.
+       * @param[out] dy_dt ODE RHS
+       * @param[in]  t time.
+       */
+      inline
+      void operator()(const std::vector<double>& y,
+                      std::vector<double>& dy_dt,
```

Comment:
I agree.  It's conventional to have the modified arguments
at the end if possible.

> On Apr 20, 2016, at 11:50 AM, Daniel Lee <notifications@github.com> wrote:
> 
> In stan/math/rev/mat/functor/ode_model.hpp:
> 
> > +          theta_(theta),
> > +          x_(x),
> > +          x_int_(x_int),
> > +          msgs_(msgs)
> > +      {}
> > +
> > +      /**
> > +       * Calculate the RHS of the ODE
> > +       *
> > +       * @param[in] y state of the ode system at time t.
> > +       * @param[out] dy_dt ODE RHS
> > +       * @param[in]  t time.
> > +       */
> > +      inline
> > +      void operator()(const std::vector<double>& y,
> > +                      std::vector<double>& dy_dt,
> 
> We usually move the parameters to an argument that are used for output to the end of the list. Could you change the order here?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>